core/app/components/pubs/PubEditor/actions.ts

```diff
@@ -16,30 +16,45 @@ import { ApiError, createPubRecursiveNew } from "~/lib/server";
 import { autoRevalidate } from "~/lib/server/cache/autoRevalidate";
 import { findCommunityBySlug } from "~/lib/server/community";
 import { defineServerAction } from "~/lib/server/defineServerAction";
-import { userHasPermissionToForm } from "~/lib/server/form";
+import { addMemberToForm, userHasPermissionToForm } from "~/lib/server/form";
 import { updatePub as _updatePub } from "~/lib/server/pub";
 
+type CreatePubRecursiveProps = Parameters<typeof createPubRecursiveNew>[0];
+
 export const createPubRecursive = defineServerAction(async function createPubRecursive(
-	...[props]: Parameters<typeof createPubRecursiveNew>
+	props: CreatePubRecursiveProps & { formSlug?: string; addUserToForm?: boolean }
 ) {
+	const { formSlug, addUserToForm, ...createPubProps } = props;
 	const loginData = await getLoginData();
 
 	if (!loginData || !loginData.user) {
 		return ApiError.NOT_LOGGED_IN;
 	}
 	const { user } = loginData;
 
-	const authorized = await userCan(
+	const canCreatePub = await userCan(
 		Capabilities.createPub,
 		{ type: MembershipType.community, communityId: props.communityId },
 		user.id
 	);
+	const canCreateFromForm = formSlug
+		? await userHasPermissionToForm({ formSlug, userId: loginData.user.id })
+		: false;
 
-	if (!authorized) {
+	if (!canCreatePub && !canCreateFromForm) {
 		return ApiError.UNAUTHORIZED;
 	}
 	try {
-		await createPubRecursiveNew(props);
+		const createdPub = await createPubRecursiveNew(createPubProps);
+
+		if (addUserToForm && formSlug) {
+			await addMemberToForm({
+				communityId: props.communityId,
+				userId: user.id,
+				slug: formSlug,
+				pubId: createdPub.id,
+			});
+		}
```

Comment:
I'm not 100% sure this is what we want but I put it in for the case where:

* A user is invited to create a pub via `/fill` (no pub id)
* They fill out the form
* Without this, after submission, the page 404s because the user doesn't actually have access to that pub id 

I figured this might be useful so that the user could also feasibly go back and edit a pub they created, but looking for feedback here! A drawback might be that this might also trigger if the user is an admin/editor, they'd also get added to the form though they don't necessarily need to be since they already have create/edit privileges to all pubs. Could get around this by checking if they already have a greater permission, though I wasn't sure if that was worth it?

Reply:
I think this is a good solution to the case you outlined. I'm not too worried about it creating redundant permissions. It may even be desireable in some edge cases (i.e. a user has their editor permission revoked and suddenly can't use a form they were invited to).